### ISSUE TITLE
root does not have permission to access...

### ISSUE THREAD

**Issue by ehusar** [rsg98/homebridge-gpio-wpi#2]
`Making wiring-pi ... gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/7.3.0"`

I am tying to install and I get the following error above after

`npm install -g homebridge-gpio-wpi`

It does promote me a bunch of time for a root password.  I am 100% positive I am entering the root password.

Here is a little bit of that password prompt section.

[Cloning libWiringPi ... done.
patching file ./wiringpi/devLib/Makefile
Hunk #1 succeeded at 24 with fuzz 1 (offset -2 lines).
Hunk #2 succeeded at 32 with fuzz 2 (offset -2 lines).
patching file ./wiringpi/gpio/Makefile
Making libWiringPi ... done.
Making devLib ...done.
Unistalling gpio utility ... [sudo] password for nobody: 
[sudo] password for nobody: 
[sudo] password for nobody: 
done.
Making gpio utility ... done.
Installing gpio utility ... [sudo] password for nobody: 
[sudo] password for nobody: 
[sudo] password for nobody: 
failed.](url)

**Comment by mikigp** [comment]
Same for me. I tried to install wiring Pi following the guide at http://wiringpi.com/download-and-install/
Installation successful... but when i launch homebridge and try to activate a pin, homebridge crashes with this error:
```
/usr/lib/node_modules/homebridge-gpio-wpi/index.js:64
    wpi.digitalWrite(self.pin, action);
        ^

Error: digitalWrite: IsInt32(arguments['state']) === false
    at GPIOAccessory.pinAction (/usr/lib/node_modules/homebridge-gpio-wpi/index.js:64:9)
    at Timeout._onTimeout (/usr/lib/node_modules/homebridge-gpio-wpi/index.js:72:14)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
```
Any ideas?

**Comment by rsg98** [comment]
There are two separate issues at work here...

@mikigp - yours is related to a bug in my module.  This should now be fixed on this Github repo, but I haven't pushed it to NPM yet (need to have a bit more testing!)  You can pull the updated code from here, or - in theory - workaround it by removing any "duration" configuration in your config.json.

@ehusar - this error is upstream in the Wiring Pi bindings (WiringPi/WiringPi-Node#38).  I'm working on a fork of that library that fixes that and pulls directly from the original WiringPi (meaning we get access to some of the newer features as well).  In the meantime, you can workaround by installing WiringPi without the -g flag - e.g.

```
sudo -i
cd `npm root -g`
npm install wiring-pi
```

You can then install homebridge-gpio-wpi normally with the -g global flag.  I'll close this once I've finished the updated bindings.

As far as I'm aware, the original reported error about root having access to .node-gyp doesn't cause any problems with the functionality of homebridge / wiring-pi / homebridge-gpio-wpi

**Comment by rsg98** [comment]
OK - [Version 0.1.0](https://www.npmjs.com/package/homebridge-gpio-wpi) pushed to NPM.  Hopefully that fixes the issue @mikigp is seeing.

**Comment by mikigp** [comment]
I tried to update my version of homebridge-gpio-wpi typing `npm install -g homebridge-gpio-wpi`.
After the installation i get the message with the Version 0.1.0. Rebooted my Pi but..  my homebridge doesn't start. This is the error:

```
[12/31/2016, 5:18:17 PM] [Garage] Initializing GPIO accessory...
[12/31/2016, 5:18:17 PM] [Garage] Configuring accessory Garage (homebridge-gpio-wpi version 0.1.0)
/usr/lib/node_modules/homebridge-gpio-wpi/index.js:32
    if(currentPinStatus.error) {
                       ^

TypeError: Cannot read property 'error' of undefined
    at new GPIOAccessory (/usr/lib/node_modules/homebridge-gpio-wpi/index.js:32:24)
    at Server._loadAccessories (/usr/lib/node_modules/homebridge/lib/server.js:265:29)
    at Server.run (/usr/lib/node_modules/homebridge/lib/server.js:81:38)
    at module.exports (/usr/lib/node_modules/homebridge/lib/cli.js:40:10)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)

```

Another issue is that my configuration of GPIO is cleared after reboot... For example if I set GPIO.29 to OUT, after reboot it came back to IN.

**Comment by rsg98** [comment]
Just pushed a 0.1.2 that should handle that error better.  I suspect what it will now say is "Pin XX is not configured..."

It's trying to see if it can read /sys/class/gpio/gpioXX - which should be created when running `gpio export XX out` as the right user (i.e. the one you're running homebridge as)

**Comment by mikigp** [comment]
Updated... 
when i've exported the pin i didn't use the "sudo" prefix, so i heve exported the status with my usual user. Have I to create the export with the "homebridge" user?
This is the new error:
```
[12/31/2016, 5:58:01 PM] [Garage] Initializing GPIO accessory...
[12/31/2016, 5:58:01 PM] [Garage] Configuring accessory Garage (homebridge-gpio-wpi version 0.1.2)
/usr/lib/node_modules/homebridge-gpio-wpi/index.js:33
      throw new Error('Pin ' + this.pin + ' is not readable (' + currentPinStatus.error.code + ').  Did you run gpio export as the right user?');
      ^

Error: Pin 29 is not readable (undefined).  Did you run gpio export as the right user?
    at new GPIOAccessory (/usr/lib/node_modules/homebridge-gpio-wpi/index.js:33:13)
    at Server._loadAccessories (/usr/lib/node_modules/homebridge/lib/server.js:265:29)
    at Server.run (/usr/lib/node_modules/homebridge/lib/server.js:81:38)
    at module.exports (/usr/lib/node_modules/homebridge/lib/cli.js:40:10)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
```

**Comment by rsg98** [comment]
Running it without sudo is fine (/usr/bin/gpio is setuid).  The easiest way is to make sure your normal user and your "homebridge" user are members of the gpio group.  For example:

```
pi@mypi:~$ ls -lart /sys/class/gpio
total 0
drwxr-xr-x 50 root root    0 Jan  1  1970 ..
lrwxrwxrwx  1 root gpio    0 Dec 31 11:22 gpiochip0 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpiochip0
lrwxrwxrwx  1 root gpio    0 Dec 31 11:22 gpio23 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio23
lrwxrwxrwx  1 root gpio    0 Dec 31 11:22 gpio24 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio24
-rwxrwx---  1 root gpio 4096 Dec 31 16:28 unexport
drwxrwx---  2 root gpio    0 Dec 31 16:43 .
lrwxrwxrwx  1 root gpio    0 Dec 31 16:43 gpio22 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio22
-rwxrwx---  1 root gpio 4096 Dec 31 16:44 export
```

You can see all the exports are owned by "root" but have read/write/execute permissions for those in the "gpio" group.

```
pi@mypi:~$ id homebridge
uid=999(homebridge) gid=996(homebridge) groups=996(homebridge),997(gpio)
```

This shows my homebridge user is in the gpio group.  If it's not, then

```
sudo usermod -G gpio homebridge
```

Would add the gpio group to the homebridge user (note capital G, which adds groups...)  Finally, to test as the homebridge user:

```
sudo -i
su - homebridge
cat /sys/class/gpio/gpio29/direction
```

...should say `out`.

**Comment by mikigp** [comment]
i followed your instruction. all fine: my homebridge user is in the gpio group, but after typing last series of commands , I receive these messages:
```
root@raspberrypi:~# su - homebridge
No directory, logging in with HOME=/
homebridge@raspberrypi:/$ cat /sys/class/gpio/gpio29/direction
cat: /sys/class/gpio/gpio29/direction: No such file or directory
```
maybe some problem with the homebridge user folder?

**Comment by rsg98** [comment]
Shouldn't be a problem with the user folder - you can make that error go away by setting a home folder for `homebridge` user (I set mine to the path where I've got homebridge config stored):

```
usermod -d /var/homebridge homebridge
```

Can you run:

```
sudo ls -al /sys/class/gpio
```

...and post the output?

**Comment by mikigp** [comment]
Home folder set for the home bridge user... Did you mean usermod, not usermode right?
This is the output:
```
pi@raspberrypi:~ $ sudo ls -al /sys/class/gpio
total 0
drwxrwx---  2 root gpio    0 Dec 31 17:56 .
drwxr-xr-x 49 root root    0 Jan  1 13:20 ..
-rwxrwx---  1 root gpio 4096 Dec 31 17:56 export
lrwxrwxrwx  1 root gpio    0 Dec 31 17:56 gpiochip0 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpiochip0
lrwxrwxrwx  1 root gpio    0 Dec 31 17:56 gpiochip100 -> ../../devices/platform/soc/soc:virtgpio/gpio/gpiochip100
-rwxrwx---  1 root gpio 4096 Dec 31 17:56 unexport
```

**Comment by rsg98** [comment]
Sorry - yes - usermod (edited my comment for clarity)

Looks like your exported pin (GPIO29) isn't there - the exports don't survive reboots!  You would need to run a script like the example set-gpio.sh each time your Pi starts up.

Could you try

```
gpio export 29 out
```

You should then see gpio29 appear under `/sys/class/gpio` and under that, the `direction` file should - hopefully - contain `out`.

If that works., then homebridge should start.  Remember to include your version of a set-gpio.sh before starting homebridge on each restart.

**Comment by mikigp** [comment]
success! homebridge started!
This is part of the output:
```
Scan this code with your HomeKit App on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    │ 031-45-154 │     
    └────────────┘     
                       
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE :::51826
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at Server._listen2 (net.js:1262:14)
    at listen (net.js:1298:10)
    at Server.listen (net.js:1376:9)
    at EventedHTTPServer.listen (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/eventedhttp.js:60:19)
    at HAPServer.listen (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/HAPServer.js:158:20)
    at Bridge.Accessory.publish (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:496:16)
    at Server._publish (/usr/lib/node_modules/homebridge/lib/server.js:114:16)
    at Server.<anonymous> (/usr/lib/node_modules/homebridge/lib/server.js:372:14) 
```
Despite these lines my homebridge is still running....
Now I'm studying in order to modify set-gpio.sh

Thank you so much for your help... I wish you an happy new year!

**Comment by rsg98** [comment]
Good news.

That error implies there are multiple copies of homebridge running (or that the port, 51826, is being used by something else)

Have a play with set-gpio.sh, it should be fairly easy to just put the pins you want into the "PINS" array - this is just standard bash scripting, nothing fancy.

Hopefully all of this will be a lot easier soon, once I've finish the wiringPi side and done the newer version of this plugin (rsg98/homebridge-gpio-wpi2) - which includes a script to generate a set-gpio.sh from the config.json (but this won't work properly for the existing plugin).

**Comment by mikigp** [comment]
Perfect! Actually not complicated!
Can I ask you the last thing? Can you point out the path of set-gpio.sh?

**Comment by rsg98** [comment]
It should be wherever the homebridge-gpio-wpi module installed - if you installed it with the "-g" flag, run

```
npm root -g
```

That should give you the folder where your node_modules are going, then inside that, should be a homebridge-gpio-wpi/set-gpio.sh script.  Failing that, you can just download it from directly from this repo... [set-gpio.sh](https://github.com/rsg98/homebridge-gpio-wpi/blob/master/set-gpio.sh)

**Comment by mikigp** [comment]
After running the the `npm root -g` (i've installed homebridge with the -g flag), I've found the full path, that in my case is:
`/usr/lib/node_modules/homebridge-gpio-wpi`

I edited it with the numbers of the pin inserted in my config.json, but after rebooting the homebridge crashed as before..
I think to have misunderstood your suggestions... The system needs t load a script for every pin, right?

**Comment by rsg98** [comment]
One script will set up all the pins - so you only need one. Just put the PIN numbers in separated by a space.

Each time the Pi reboots, you need to run the script before starting homebridge.

Depending on how you're starting homebridge, there are different options for automating this e.g. As an ExecStartPre script if you're using systemd.

**Comment by mikigp** [comment]
I changed the "out" part of the set-gpio.sh script, removing the "-g" part (i really necessary the "down" command?):
```
#!/bin/sh

#
# Example setup script for homebridge-gpio-wpi
#
# Make sure that the user you're running homebridge as
# is a member of the 'gpio' group.
#

GPIO=/usr/local/bin/gpio

# These are the BCM_GPIO Pin Numbers!
PINS='27 28 29'

for i in $PINS
do
        $GPIO mode $i out
        $GPIO mode $i down

        $GPIO export $i out
done
```
because homebridge export files report BCM number, when homebridge search for the "GPIO" numer..

After that i tried to modify my homebridge.service in order to execute set-gpio.sh file on boot... but without success until now:
```
[Unit]
Description=Node.js HomeKit Server
After=syslog.target network-online.target

[Service]
Type=simple
User=homebridge
EnvironmentFile=/etc/default/homebridge
ExecStartPre=bash /usr/lib/node_modules/homebridge-gpio-wpi/set-gpio.sh
ExecStart=/usr/bin/homebridge $HOMEBRIDGE_OPTS
Restart=on-failure
RestartSec=10
KillMode=process

[Install]
WantedBy=multi-user.target
```
If anyone have a suggestion... ;)

P.s. after executing the script manually all works fine!

**Comment by rsg98** [comment]
Try

```
ExecStartPre=/bin/sh /usr/lib/node_modules/homebridge-gpio-wpi/set-gpio.sh
```

(Or /bin/bash, but it might not be able to find bash if there's no `PATH` set when systemd starts.  Also, if you run `systemctl status homebridge` - it should tell you the exit status of your set-gpio script on the Process line:

```
● homebridge.service - Node.js HomeKit Server
   Loaded: loaded (/etc/systemd/system/homebridge.service; enabled)
   Active: active (running) since Sat 2016-12-31 16:44:31 GMT; 1 day 1h ago
  Process: 3939 ExecStartPre=/bin/sh /usr/local/bin/set-gpio.sh (code=exited, status=0/SUCCESS)
 Main PID: 3951 (homebridge)
   CGroup: /system.slice/homebridge.service
           └─3951 homebridge
```

**Comment by mikigp** [comment]
Perfect! Now it works! This is the final code for the homebridge.service:
```
[Unit]
Description=Node.js HomeKit Server
After=syslog.target network-online.target

[Service]
Type=simple
User=homebridge
EnvironmentFile=/etc/default/homebridge
ExecStartPre=/bin/sh /usr/lib/node_modules/homebridge-gpio-wpi/set-gpio.sh
ExecStart=/usr/bin/homebridge $HOMEBRIDGE_OPTS
Restart=on-failure
RestartSec=10
KillMode=process

[Install]
WantedBy=multi-user.target
```
Thank you again for the help-to-the-newbie!

**Comment by mikigp** [comment]
Solved: after testing, the problem was in the pin definition: to avoid any confusion, I changed all the references into the files referring to the BCM numbering table, as rsg98 recommended.

Another issue...
It seems the software part of the project is now "ok", i've tried to test the hardware part, but when i activate from my apple device the pin 29 (that acts as a button) through the Home App nothing happens on gpio board... What's wrong now? From the gpio utility all works fine, but not from homebridge.

**Comment by rsg98** [comment]
@mikigp glad you got it figured out.  If it appears to work in Homebridge, but doesn't actually do anything, then it's generally either permissions (which is why I make it throw an Error if it can't read the pin exports) or the overly-complicated number of different pin numbers!

(It doesn't help that `gpio` takes the WiringPi Pin number in all modes (unless you specify `-g`) apart from `export`, where it takes the BCM number!)

**Comment by mikigp** [comment]
I received now my 4 channel relay to complete my setup for my Garage Door. There is another strange thing... When connected to the Pi, all the relay are excited: when i activate a pin through homebridge the relay witch off. In other words an inverted behavior... It should not be the opposite?

**Comment by rsg98** [comment]
Others have said the same thing with some of the relay boards - so you can set `"inverted" : "true"` in your config file against each of the GPIO pins you want to behave the "wrong" way around.  

I've not got one of those boards, so not sure if there is a better way of solving it electronically, rather in software - but the inverted bodge seems to work!

**Comment by rsg98** [comment]
homebridge-gpio-wpi2 uses new node-wiring-pi library, which supports -g global mode install.  Migrate to homebridge-gpio-wpi2.
